common/beerocks/bcl/include/bcl/beerocks_event_loop.h

```diff
@@ -0,0 +1,165 @@
+ * event loop is at the highest level of control within the program.
+ */
+template <typename T> class EventLoop {
+public:
+    /**
+     * The type of the event source (e.g. file descriptor).
+     */
+    using EventType = T;
+
+    /**
+     * The type of the event loop.
+     */
+    using EventLoopType = EventLoop<T>;
+
+    /**
+     * @brief Event handler function definition.
+     *
+     * Parameters to the event handler function are:
+     * - The shared pointer to a socket resource where the event was originated at. 
+     * Event handler can perform, for example, accept, read, write and close operations on the socket.
+     * - The event loop where it was caught. Event handler can install new handlers, remove existing
+     * handlers and even ask for loop termination.
+     * 
+     * @returns True on success or false otherwise
+     */
+    using EventHandler = std::function<bool(EventType fd, EventLoopType &loop)>;
+
+    /**
+     * Set of event handler functions, one function to handle each possible event happened.
+     * Handlers are grouped into a struct to facilitate passing them as a single parameter to the install and
+     * remove handlers functions of the event loop.
+     * Event handlers are optional and if not set for a given event, that event will be silently
+     * ignored.
+     */
+    struct EventHandlers {
+        /**
+         * Hook method that is called back by the event loop to handle read events.
+         * Read events are dispatched when the socket is ready for a read operation (a read
+         * operation will not block).
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_read;
+
+        /**
+         * Hook method that is called back by the event loop to handle write events.
+         * Write events are dispatched when the socket is ready for a write operation (a write
+         * operation will not block).
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_write;
+
+        /**
+         * Hook method that is called back by the event loop to handle timeout events.
+         * Timeout events are dispatched when timeout expires while waiting for the socket to
+         * be ready for a read or write operation.
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_timeout;
+
+        /**
+         * Hook method that is called back by the event loop to handle disconnect events.
+         * Disconnect events are dispatched when the remote socket is closed.
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_disconnect;
```

Comment:
I like it that on_disconnect is separate from on_error! They will usually be the same, but it's more convenient to distinguish in the eventloop itself rather than afterwards.